tikv/raftstore/read.go

```
@@ -0,0 +1,100 @@
+package raftstore
+
+import (
+	"fmt"
+	stdatomic "sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/raft_cmdpb"
+	"github.com/uber-go/atomic"
+)
+
+type LeaderChecker interface {
+	IsLeader(ctx *kvrpcpb.Context, router RaftstoreRouter) error
+}
+
+type leaderChecker struct {
+	invalid          atomic.Bool
+	peerID           atomic.Uint64
+	term             atomic.Uint64
+	appliedIndexTerm atomic.Uint64
+	leaderLease      unsafe.Pointer // *RemoteLease
+	region           unsafe.Pointer // *metapb.Region
+}
+
+func (c *leaderChecker) IsLeader(ctx *kvrpcpb.Context, router RaftstoreRouter) error {
+	snapTime := time.Now()
+	isExpired, err := c.isExpired(ctx, &snapTime)
+	if err != nil {
+		return err
+	}
+	if !isExpired {
+		return nil
+	}
+
+	cb := NewCallback()
+	req := new(raft_cmdpb.Request)
+	req.CmdType = raft_cmdpb.CmdType_Snap
+	header := &raft_cmdpb.RaftRequestHeader{
+		RegionId:    ctx.RegionId,
+		Peer:        ctx.Peer,
+		RegionEpoch: ctx.RegionEpoch,
+		Term:        ctx.Term,
+		SyncLog:     ctx.SyncLog,
+	}
+	cmd := &raft_cmdpb.RaftCmdRequest{
+		Header:   header,
+		Requests: []*raft_cmdpb.Request{req},
+	}
+
+	err = router.SendCommand(cmd, cb)
+	if err != nil {
+		return err
+	}
+
+	cb.wg.Wait()
+
+	if cb.resp.Header.Error != nil {
+		return &RaftError{e: cb.resp.Header.Error}
+	}
+	return nil
+}
+
+func (c *leaderChecker) isExpired(ctx *kvrpcpb.Context, snapTime *time.Time) (bool, error) {
+	if c.invalid.Load() {
+		return false, nil
```

Can we return an error when it is invalid?